massa-execution-worker/src/speculative_ledger.rs

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2022 MASSA LABS <info@massa.net>
+
+//! The speculative ledger represents, in a compressed way,
+//! the state of the ledger at an arbitrary execution slot.
+//! It never actually writes to the consensus state
+//! but keeps track of the changes that were applied to it since its creation.
+
+use massa_execution_exports::ExecutionError;
+use massa_hash::hash::Hash;
+use massa_ledger::{Applicable, FinalLedger, LedgerChanges};
+use massa_models::{Address, Amount};
+use std::sync::{Arc, RwLock};
+
+/// The SpeculativeLedger contains an exclusive reference to the final ledger (read-only),
+/// a list of existing changes that happened o the ledger since its finality,
+/// as well as an extra list of "added" changes.
+/// The SpeculativeLedger makes it possible to transparently manipulate a virtual ledger
+/// that takes into account all those ledger changes and allows adding more
+/// while keeping track of all the newly added changes, and never writing in the final ledger.
+pub struct SpeculativeLedger {
+    /// Exclusive access to the final ledger. For reading only.
```

Comment:
Why do you write "exclusive" here(and I think elsewhere?). Those are all shared reference right?

Reply:
corrected